src/balance.rs

```diff
@@ -26,41 +42,186 @@ impl<T: Config> From<Error<T>> for StatusCode {
+    ) -> DispatchResult {
+        // TODO (neutrinoks): Enable this feature to ensure, that balance is available!
+        // self.ensure_can_withdraw(account, balance)?;
+
+        let mut cheques = self.cheques.borrow_mut();
+        if let Some(src_cheque) = cheques.get_mut(account) {
+            *src_cheque += *balance;
+        } else {
+            cheques.insert(account.clone(), *balance);
+        }
+
+        if let Some(src_cheque) = self.initial_state.get_mut(account) {
+            *src_cheque += *balance;
+        } else {
+            self.initial_state.insert(account.clone(), *balance);
+        }
+
+        Ok(())
+    }
+
+    /// Executes the true transactions on the blockchain/substrate side after execution of
+    /// Move-script.
+    pub fn apply_transactions(&self) -> DispatchResult {
+        use core::cmp::Ordering;
+
+        let zero = BalanceOf::<T>::zero();
+
+        self.cmp_with_initial_state()?;
+
+        let cheques = self.cheques.borrow();
+        let mut purse = NegativeImbalanceOf::<T>::zero();
+        let mut depts = Vec::<(&AccountIdOf<T>, BalanceOf<T>)>::new();
+
+        // Calculate balance differences and withdraw negative ones from user's accounts.
+        for (account, balance) in cheques.iter() {
+            let true_balance = self.initial_state.get(account).unwrap_or(&zero);
+            match (*true_balance).cmp(balance) {
+                Ordering::Greater => {
+                    let dept = *true_balance - *balance;
+                    let imbalance = T::Currency::withdraw(
+                        account,
+                        dept,
+                        WithdrawReasons::TRANSFER,
+                        ExistenceRequirement::AllowDeath,
+                    )?;
+                    purse = purse.merge(imbalance);
+                }
+                Ordering::Less => {
+                    let dept = *balance - *true_balance;
+                    depts.push((account, dept));
+                }
+                Ordering::Equal => {}
+            }
+        }
+
+        // Now deposit depts from purse to new owners.
+        for (account, balance) in depts.into_iter() {
+            let imbalance = purse.extract(balance);
+            T::Currency::resolve_creating(account, imbalance);
+        }
+
+        Ok(())
+    }
+
+    /// Ensures that user can withdraw that given amount of money, which eventually will be used
+    /// within the Move-script execution.
+    #[allow(dead_code)]
+    fn ensure_can_withdraw(
+        &self,
+        account: &AccountIdOf<T>,
+        amount: &BalanceOf<T>,
+    ) -> DispatchResult {
+        if *amount < T::Currency::free_balance(account) {
+            Ok(())
+        } else {
+            Err(Error::<T>::InsufficientBalance.into())
+        }
+    }
+
+    /// Does a state checking on initial state of cheques with current state.
+    fn cmp_with_initial_state(&self) -> DispatchResult {
+        use sp_std::cell::Ref;
```

Comment:
Do you prefer it this way or can we move it to the top of the file?

Reply:
Done